kore/src/Kore/Rewrite/Axiom/Matcher.hs

```diff
@@ -81,696 +68,401 @@ import Kore.Internal.Predicate (
-    field @"substitution" <>= subst
-
-    -- Isolate the deferred pairs which depend on the variable.
-    -- After substitution, the dependent pairs go to the front of the queue.
-    MatcherState{deferred} <- Monad.State.get
-    let (indep, dep) = Seq.partition isIndependent deferred
-    field @"deferred" .= indep
-
-    -- Push the dependent deferred pairs to the front of the queue.
-    traverse_ push dep
-    -- Apply the substitution to the queued pairs.
-    field @"queued" . transformQueue Lens.mapped %= substitute2
-
-    -- Apply the substitution to the accumulated matching solution.
-    field @"substitution" . Lens.mapped %= substitute1
-    field @"predicate" %= MultiAnd.map (Substitute.substitute subst)
-
-    return ()
-  where
-    variable = inject sVariable
-    Variable{variableName} = variable
-    isIndependent = not . any (hasFreeVariable variableName)
-    subst = Map.singleton variableName termLike
-    substitute2 :: Constraint variable -> Constraint variable
-    substitute2 (Constraint pair) = Constraint $ fmap substitute1 pair
-    substitute1 = substituteTermLike subst
-
-transformQueue ::
-    Functor f =>
-    Ord a =>
-    ((a -> f a) -> [a] -> f [a]) ->
-    (a -> f a) ->
-    Set a ->
-    f (Set a)
-transformQueue trans f x = Set.fromList <$> trans f (Set.toAscList x)
-
-substituteTermLike ::
-    MatchingVariable variable =>
-    Map (SomeVariableName variable) (TermLike variable) ->
-    TermLike variable ->
-    TermLike variable
-substituteTermLike subst = renormalizeBuiltins . Substitute.substitute subst
-
-occursCheck ::
-    (MatchingVariable variable, Monad simplifier) =>
-    SomeVariable variable ->
-    TermLike variable ->
-    MaybeT (MatcherT variable simplifier) ()
-occursCheck Variable{variableName} termLike =
-    (Monad.guard . not) (hasFreeVariable variableName termLike)
-
-definedTerm ::
-    MatchingVariable variable =>
-    MonadState (MatcherState variable) matcher =>
-    MonadReader (SideCondition variable) matcher =>
-    TermLike variable ->
-    matcher ()
-definedTerm termLike =
-    ask >>= definedTermWorker
-  where
-    definedTermWorker sideCondition
-        | SideCondition.isDefined sideCondition termLike =
-            return ()
-        | otherwise = field @"predicate" <>= definedTermLike
+        Simplifier (Either Text (MatchResult RewritingVariableName))
+    bind var term =
+        let varName = variableName var
+            freeVars = FreeVariables.toNames (freeVariables term)
+         in if not $ Set.disjoint freeVars boundSet
+                then failMatch "bound variable would escape binder"
+                else case Map.lookup varName subst of
+                    Nothing -> patternMatch' sideCondition rest deferred (isTermDefined var term) (Map.insert (variableName var) term subst)
+                    Just binding -> if binding == term then patternMatch' sideCondition rest deferred predicate subst else failMatch "nonlinear matching fails equality test"
```

Comment:
These calls to `patternMatch'` are recursive, right? Is there no way to avoid this, or to refactor it in such a way so that the recursive calls are more explicit/at the top?

Reply:
Very good point, actually. This is in fact a state-monadic computation; each case above ultimately has to `failMatch` or call `patternMatch'` recursively.
The old code used `RWST` to separate the side condition, which is probably not necessary, but it would greatly improve readability if this code was using a state monad. Of course this is a big refactoring.

Reply:
I'm not hugely convinced it makes the code that much more readable; The recursive calls to patternMatch' are already factored out into methods that are pretty clear what their meaning is. It might help things somewhat, but probably not hugely, and it would be a pretty substantial code change. If it turns out to add overhead (and I don't see how it possibly wouldn't), then it wouldn't even be a change worth making since our goal in this PR is to make pattern matching as fast as possible.

I'm willing to give it a try after all other code suggestions on this PR have been resolved if you still think it's worth trying, though.